### PR TITLE
Restrict usages of StoreModule even further.

### PR DESCRIPTION
**Pull request by sjrd**
~~Alternative to #5057. The first commit is the same as the one in #5057.~~

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala [hunk 1]
```diff
@@ -592,14 +590,30 @@ private final class ClassDefChecker(classDef: ClassDef,
...
-     */
-    if (classDef.kind.hasModuleAccessor)
-      trees.filter(!_.isInstanceOf[StoreModule])
-    else
+  private def handleStoreModulesAfterSuperCtorCall(trees: List[Tree])(
+      implicit ctx: ErrorContext): List[Tree] = {
+
+    if (classDef.kind == ClassKind.ModuleClass) {
+      if (postOptimizer) {
+        /* If the super constructor call was inlined, the StoreModule can be anywhere.
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:598]: Ah, I think this answers my question about it being implicit :)

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala [hunk 2]
```diff
@@ -331,9 +331,7 @@ private final class ClassDefChecker(classDef: ClassDef,
     val envJustBeforeSuper = checkBlockStats(body.beforeSuper, startEnv)
     checkTreeOrSpreads(body.superCall.args, envJustBeforeSuper)
     val envJustAfterSuper = envJustBeforeSuper.withThisType(instanceThisType)
-    val afterSuperStoreModulesHandled =
-      handleStoreModulesAfterSuperCtorCall(body.afterSuper)
-    checkBlockStats(afterSuperStoreModulesHandled, envJustAfterSuper)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala]: As I commented on #5057, I don't think we should do this (but start emitting store module for JS modules as well).

File: linker/shared/src/test/scala/org/scalajs/linker/checker/ClassDefCheckerTest.scala
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/checker/ClassDefCheckerTest.scala:1]: Add a test that tests a `StoreModule` at the top level after the the super ctor call (but not immediately after) is not allowed.

**Review by gzm0 — CHANGES_REQUESTED** [review]

**Review by gzm0 — APPROVED** [review]

**Comment by sjrd** [comment]
Fixed scalastyle 🤦‍♂️

**Pull request merged**